File: server_py/routers/used.py
from fastapi import APIRouter, Depends, HTTPException
from sqlmodel import Session, select
from typing import List, Optional
from db import get_session
from models import UsedItem, User
from .auth import get_current_user, get_current_admin
import uuid
import json
from datetime import datetime

router = APIRouter()

@router.get("/", response_model=dict)
@router.get("", response_model=dict)
async def get_used_items(
    type: Optional[str] = None,
    category: Optional[str] = None,
    condition: Optional[str] = None,
    status: Optional[str] = "published",
    search: Optional[str] = None,
    page: int = 1,
    page_size: int = 20,
    session: Session = Depends(get_session)
):
    query = select(UsedItem)
    if status != "all":
        query = query.where(UsedItem.status == status)
    
    if type:
        query = query.where(UsedItem.type == type)
    if category:
        query = query.where(UsedItem.category == category)
    if condition:
        query = query.where(UsedItem.condition == condition)
    
    if search:
        from sqlalchemy import or_, func
        keywords = search.strip().split()
        for kw in keywords:
            search_term = f"%{kw}%"
            query = query.where(or_(
                func.coalesce(UsedItem.brand, "").ilike(search_term),
                func.coalesce(UsedItem.model, "").ilike(search_term),
                func.coalesce(UsedItem.description, "").ilike(search_term)
            ))
        
    # Count total
    from sqlalchemy import func
    count_query = select(func.count()).select_from(UsedItem)
    if status != "all":
        count_query = count_query.where(UsedItem.status == status)
        
    if type: count_query = count_query.where(UsedItem.type == type)
    if category: count_query = count_query.where(UsedItem.category == category)
    if condition: count_query = count_query.where(UsedItem.condition == condition)
    if search:
        from sqlalchemy import or_, func
        keywords = search.strip().split()
        for kw in keywords:
            search_term = f"%{kw}%"
            count_query = count_query.where(or_(
                func.coalesce(UsedItem.brand, "").ilike(search_term),
                func.coalesce(UsedItem.model, "").ilike(search_term),
                func.coalesce(UsedItem.description, "").ilike(search_term)
            ))
    total = session.scalar(count_query)

    offset = (page - 1) * page_size
    items = session.exec(query.order_by(UsedItem.createdAt.desc()).offset(offset).limit(page_size)).all()
    
    return {
        "items": items,
        "total": total,
        "page": page,
        "page_size": page_size
    }

@router.get("/admin", response_model=dict)
async def get_admin_used_items(
    page: int = 1,
    page_size: int = 20,
    category: Optional[str] = None,
    brand: Optional[str] = None,
    search: Optional[str] = None,
    session: Session = Depends(get_session),
    admin: User = Depends(get_current_admin)
):
    from sqlalchemy import func, or_
    query = select(UsedItem)
    if category and category != 'all':
        query = query.where(UsedItem.category == category)
    if brand and brand != 'all':
        query = query.where(UsedItem.brand == brand)
    if search:
        keywords = search.strip().split()
        for kw in keywords:
            search_term = f"%{kw}%"
            query = query.where(or_(
                func.coalesce(UsedItem.brand, "").ilike(search_term),
                func.coalesce(UsedItem.model, "").ilike(search_term),
                func.coalesce(UsedItem.description, "").ilike(search_term)
            ))

    count_query = select(func.count()).select_from(UsedItem)
    if category and category != 'all': count_query = count_query.where(UsedItem.category == category)
    if brand and brand != 'all': count_query = count_query.where(UsedItem.brand == brand)
    if search:
        keywords = search.strip().split()
        for kw in keywords:
            search_term = f"%{kw}%"
            count_query = count_query.where(or_(
                func.coalesce(UsedItem.brand, "").ilike(search_term),
                func.coalesce(UsedItem.model, "").ilike(search_term)
            ))

    total = session.scalar(count_query)
    offset = (page - 1) * page_size
    items = session.exec(query.order_by(UsedItem.createdAt.desc()).offset(offset).limit(page_size)).all()
    
    return {
        "items": items,
        "total": total,
        "page": page,
        "page_size": page_size
    }

@router.post("/")
@router.post("")
async def create_used_item(
    item_data: dict, 
    session: Session = Depends(get_session),
    user: User = Depends(get_current_user)
):
    new_item = UsedItem(
        id=str(uuid.uuid4()),
        type=item_data.get("type", "personal"),
        sellerId=user.id,
        sellerName=user.username,
        contact=item_data.get("contact", ""),
        category=item_data.get("category"),
        brand=item_data.get("brand"),
        model=item_data.get("model"),
        price=item_data.get("price"),
        originalPrice=item_data.get("originalPrice"),
        condition=item_data.get("condition"),
        images=json.dumps(item_data.get("images", [])),
        description=item_data.get("description"),
        status=item_data.get("status", "pending") if user.role == "admin" else "pending",
        inspectionReport=json.dumps(item_data.get("inspectionReport")) if item_data.get("inspectionReport") else "null"
    )
    session.add(new_item)
    session.commit()
    session.refresh(new_item)
    return new_item

@router.put("/{item_id}")
async def update_used_item(
    item_id: str,
    item_data: dict,
    session: Session = Depends(get_session),
    user: User = Depends(get_current_user)
):
    item = session.get(UsedItem, item_id)
    if not item:
        raise HTTPException(status_code=404, detail="商品未找到")
    
    # Check ownership or admin
    if item.sellerId != user.id and user.role != "admin":
        raise HTTPException(status_code=403, detail="无权操作")
    
    for key, value in item_data.items():
        if hasattr(item, key):
            if key == "images" and isinstance(value, list):
                value = json.dumps(value)
            if key == "inspectionReport" and value is not None:
                value = json.dumps(value)
            setattr(item, key, value)
            
    session.add(item)
    session.commit()
    session.refresh(item)
    return item

@router.delete("/{item_id}")
async def delete_used_item(
    item_id: str,
    session: Session = Depends(get_session),
    user: User = Depends(get_current_user)
):
    item = session.get(UsedItem, item_id)
    if not item:
        raise HTTPException(status_code=404, detail="商品未找到")
    
    if item.sellerId != user.id and user.role != "admin":
        raise HTTPException(status_code=403, detail="无权操作")
    
    session.delete(item)
    session.commit()
    return {"message": "商品已删除"}

@router.post("/{item_id}/mark-sold")
async def mark_sold(
    item_id: str,
    session: Session = Depends(get_session),
    user: User = Depends(get_current_user)
):
    item = session.get(UsedItem, item_id)
    if not item:
        raise HTTPException(status_code=404, detail="商品未找到")
    
    if item.sellerId != user.id and user.role != "admin":
        raise HTTPException(status_code=403, detail="无权操作")
    
    item.status = "sold"
    session.add(item)
    session.commit()
    return {"message": "已标记为已售"}

File: server/routes/used.ts
import express from 'express';
import { db } from '../db';
import { authenticate, authorize } from '../middleware/auth';
import crypto from 'crypto';

const router = express.Router();

// Get published used items (Public)
router.get('/', async (req, res) => {
    try {
        const rows = db.prepare("SELECT * FROM used_items WHERE status = 'published' ORDER BY createdAt DESC").all();
        const items = rows.map((row: any) => {
            let images = [];
            let report = null;
            try {
                images = row.images ? JSON.parse(row.images) : [];
                report = row.inspectionReport ? JSON.parse(row.inspectionReport) : null;
            } catch (e) {
                console.error(`Failed to parse JSON for used item ${row.id}:`, e);
            }
            return { ...row, images, inspectionReport: report };
        });
        res.json(items);
    } catch (error) {
        console.error('Failed to fetch used items:', error);
        res.status(500).json({ error: 'Failed to fetch used items', details: (error as Error).message });
    }
});

// Admin/Streamer: Get all items
router.get('/admin', authenticate, authorize(['admin', 'streamer']), async (req, res) => {
    try {
        const rows = db.prepare('SELECT * FROM used_items ORDER BY createdAt DESC').all();
        const items = rows.map((row: any) => {
            let images = [];
            let report = null;
            try {
                images = row.images ? JSON.parse(row.images) : [];
                report = row.inspectionReport ? JSON.parse(row.inspectionReport) : null;
            } catch (e) {
                console.error(`Failed to parse JSON for used item ${row.id}:`, e);
            }
            return { ...row, images, inspectionReport: report };
        });
        res.json(items);
    } catch (error) {
        console.error('Failed to fetch admin used items:', error);
        res.status(500).json({ error: 'Failed to fetch all used items', details: (error as Error).message });
    }
});

// Create/Update item
router.post('/', authenticate, authorize(['admin', 'streamer']), async (req, res) => {
    const item = req.body;
    try {
        const id = item.id || item._id;
        const images = JSON.stringify(item.images || []);
        const report = item.inspectionReport ? JSON.stringify(item.inspectionReport) : null;
        const now = new Date().toISOString();

        // Mandatory columns with defaults if missing
        const price = item.price || 0;
        const condition = item.condition || '99新';
        const description = item.description || '';
        const status = item.status || 'pending';

        if (id) {
            db.prepare(`
                UPDATE used_items 
                SET type = ?, sellerId = ?, sellerName = ?, contact = ?, category = ?, brand = ?, model = ?, price = ?, originalPrice = ?, condition = ?, images = ?, description = ?, status = ?, inspectionReport = ?, soldAt = ?
                WHERE id = ?
            `).run(item.type, item.sellerId, item.sellerName, item.contact, item.category, item.brand, item.model, price, item.originalPrice, condition, images, description, status, report, item.soldAt || null, id);
            res.json(db.prepare('SELECT * FROM used_items WHERE id = ?').get(id));
        } else {
            const newId = crypto.randomUUID();
            db.prepare(`
                INSERT INTO used_items (id, type, sellerId, sellerName, contact, category, brand, model, price, originalPrice, condition, images, description, status, inspectionReport, createdAt)
                VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
            `).run(newId, item.type, item.sellerId, item.sellerName, item.contact, item.category, item.brand, item.model, price, item.originalPrice, condition, images, description, status, report, now);
            res.json(db.prepare('SELECT * FROM used_items WHERE id = ?').get(newId));
        }
    } catch (error) {
        console.error('Failed to save used item:', error);
        res.status(500).json({ error: 'Failed to save used item' });
    }
});

// Delete item (Admin only)
router.delete('/:id', authenticate, authorize(['admin']), async (req, res) => {
    try {
        db.prepare('DELETE FROM used_items WHERE id = ?').run(req.params.id);
        res.json({ success: true });
    } catch (error) {
        res.status(500).json({ error: 'Failed to delete used item' });
    }
});

export default router;
